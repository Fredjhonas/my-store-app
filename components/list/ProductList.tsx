import { EMPTY_PRODUCT_LIST } from '@/constants/Messages';
import { Product } from '@/models/ProductModel';
import { Dimensions, FlatList, ImageStyle, TouchableOpacity, View, ViewStyle } from 'react-native';
import { ThemedText } from '../ThemedText';
import ImageLoader from '../ui/ImageLoader';

type ProductListProps = {
  products: Product[];
  isEmpty: boolean;
  isRefreshing: boolean;
  onRefresh: () => void;
  onPressItem: (itemId: number) => void;
};

const SCREEN_HEIGHT = Dimensions.get('window').height;

const ProductList = ({
  products,
  isEmpty,
  isRefreshing,
  onRefresh,
  onPressItem,
}: ProductListProps) => {
  const renderEmptyComponent = () => {
    if (!isEmpty) return null;

    return (
      <View style={$emptyContainer}>
        <ThemedText type="subtitle">{EMPTY_PRODUCT_LIST}</ThemedText>
      </View>
    );
  };

  const renderItem = ({ item }: { item: Product }) => {
    return (
      <TouchableOpacity style={$itemStyle} onPress={() => onPressItem(item.id)}>
        <ImageLoader imageUrl={item.image} style={$imageStyle} />
        <ThemedText
          style={{ fontWeight: 'bold', fontSize: 16 }}
          numberOfLines={2}
          ellipsizeMode="tail"
        >
          {item.title}
        </ThemedText>
        <ThemedText type="subtitle">${item.price}</ThemedText>
      </TouchableOpacity>
    );
  };

  return (
    <FlatList
      testID="product-list"
      data={products}
      keyExtractor={(item) => item.id.toString()}
      renderItem={renderItem}
      showsVerticalScrollIndicator={false}
      columnWrapperStyle={{ justifyContent: 'space-between' }}
      numColumns={2}
      contentContainerStyle={$listContainer}
      ListEmptyComponent={renderEmptyComponent}
      onRefresh={onRefresh}
      refreshing={isRefreshing}
      ListHeaderComponent={
        <ThemedText type="title" style={{ textAlign: 'center', marginVertical: 10 }}>
          Listado de Productos
        </ThemedText>
      }
    />
  );
};

export default ProductList;

const $itemStyle: ViewStyle = {
  padding: 20,
  backgroundColor: 'white',
  margin: 10,
  borderRadius: 10,
  width: '45%',
  gap: 10,
};

const $imageStyle: ImageStyle = {
  width: '100%',
  height: SCREEN_HEIGHT * 0.2,
  borderRadius: 10,
};

const $listContainer: ViewStyle = {
  padding: 10,
  flexGrow: 1,
  paddingBottom: 20,
};

const $emptyContainer: ViewStyle = {
  flex: 1,
  justifyContent: 'center',
  alignItems: 'center',
};
