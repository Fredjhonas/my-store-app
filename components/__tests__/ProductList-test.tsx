import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import ProductList from '../list/ProductList';

describe('ProductList Component', () => {
  const mockProducts = [
    {
      id: 1,
      title: 'Product 1',
      price: 10,
      image: 'image1.png',
      description: 'Description 1',
      category: 'Category 1',
      rating: { rate: 4.5, count: 100 },
    },
    {
      id: 2,
      title: 'Product 2',
      price: 20,
      image: 'image2.png',
      description: 'Description 2',
      category: 'Category 2',
      rating: { rate: 4.0, count: 200 },
    },
  ];

  it('renders correctly with products', () => {
    const { getByText } = render(
      <ProductList
        products={mockProducts}
        isEmpty={false}
        isRefreshing={false}
        onRefresh={jest.fn()}
        onPressItem={jest.fn()}
      />
    );

    mockProducts.forEach((product) => {
      expect(getByText(product.title)).toBeTruthy();
      expect(getByText(`$${product.price}`)).toBeTruthy();
    });
  });

  it('renders empty state when no products are provided', () => {
    const { getByText } = render(
      <ProductList
        products={[]}
        isEmpty={true}
        isRefreshing={false}
        onRefresh={jest.fn()}
        onPressItem={jest.fn()}
      />
    );

    expect(getByText('No hay productos disponibles')).toBeTruthy();
  });

  it('calls onRefresh when the list is pulled down', () => {
    const mockOnRefresh = jest.fn();
    const { getByTestId } = render(
      <ProductList
        products={mockProducts}
        isEmpty={false}
        isRefreshing={false}
        onRefresh={mockOnRefresh}
        onPressItem={jest.fn()}
      />
    );

    fireEvent(getByTestId('product-list'), 'refresh');
    expect(mockOnRefresh).toHaveBeenCalledTimes(1);
  });

  it('calls onPressItem when a product is pressed', () => {
    const mockOnPressItem = jest.fn();
    const { getByText } = render(
      <ProductList
        products={mockProducts}
        isEmpty={false}
        isRefreshing={false}
        onRefresh={jest.fn()}
        onPressItem={mockOnPressItem}
      />
    );

    fireEvent.press(getByText('Product 1'));
    expect(mockOnPressItem).toHaveBeenCalledWith(1);
  });

  it('matches snapshot', () => {
    const { toJSON } = render(
      <ProductList
        products={mockProducts}
        isEmpty={false}
        isRefreshing={false}
        onRefresh={jest.fn()}
        onPressItem={jest.fn()}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
