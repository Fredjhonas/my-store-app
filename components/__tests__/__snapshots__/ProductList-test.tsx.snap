// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList Component matches snapshot 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  ListHeaderComponent={
    <ThemedText
      style={
        {
          "marginVertical": 10,
          "textAlign": "center",
        }
      }
      type="title"
    >
      Listado de Productos
    </ThemedText>
  }
  contentContainerStyle={
    {
      "flexGrow": 1,
      "padding": 10,
      "paddingBottom": 20,
    }
  }
  data={
    [
      {
        "category": "Category 1",
        "description": "Description 1",
        "id": 1,
        "image": "image1.png",
        "price": 10,
        "rating": {
          "count": 100,
          "rate": 4.5,
        },
        "title": "Product 1",
      },
      {
        "category": "Category 2",
        "description": "Description 2",
        "id": 2,
        "image": "image2.png",
        "price": 20,
        "rating": {
          "count": 200,
          "rate": 4,
        },
        "title": "Product 2",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onRefresh={[MockFunction]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[MockFunction]}
      refreshing={false}
    />
  }
  refreshing={false}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  showsVerticalScrollIndicator={false}
  stickyHeaderIndices={[]}
  testID="product-list"
  viewabilityConfigCallbackPairs={[]}
>
  <RCTRefreshControl />
  <View>
    <View
      collapsable={false}
      onLayout={[Function]}
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            undefined,
            {
              "fontSize": 32,
              "fontWeight": "bold",
              "lineHeight": 32,
            },
            undefined,
            undefined,
            undefined,
            {
              "marginVertical": 10,
              "textAlign": "center",
            },
          ]
        }
      >
        Listado de Productos
      </Text>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "justifyContent": "space-between",
            },
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "gap": 10,
              "margin": 10,
              "opacity": 1,
              "padding": 20,
              "width": "45%",
            }
          }
        >
          <Image
            resizeMode="contain"
            source={
              {
                "uri": "image1.png",
              }
            }
            style={
              {
                "borderRadius": 10,
                "height": 266.8,
                "width": "100%",
              }
            }
          />
          <Text
            ellipsizeMode="tail"
            numberOfLines={2}
            style={
              [
                {
                  "color": "#11181C",
                },
                {
                  "fontSize": 16,
                  "lineHeight": 24,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                {
                  "fontSize": 16,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            Product 1
          </Text>
          <Text
            style={
              [
                {
                  "color": "#11181C",
                },
                undefined,
                undefined,
                undefined,
                {
                  "fontSize": 20,
                  "fontWeight": "bold",
                },
                undefined,
                undefined,
              ]
            }
          >
            $
            10
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "gap": 10,
              "margin": 10,
              "opacity": 1,
              "padding": 20,
              "width": "45%",
            }
          }
        >
          <Image
            resizeMode="contain"
            source={
              {
                "uri": "image2.png",
              }
            }
            style={
              {
                "borderRadius": 10,
                "height": 266.8,
                "width": "100%",
              }
            }
          />
          <Text
            ellipsizeMode="tail"
            numberOfLines={2}
            style={
              [
                {
                  "color": "#11181C",
                },
                {
                  "fontSize": 16,
                  "lineHeight": 24,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                {
                  "fontSize": 16,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            Product 2
          </Text>
          <Text
            style={
              [
                {
                  "color": "#11181C",
                },
                undefined,
                undefined,
                undefined,
                {
                  "fontSize": 20,
                  "fontWeight": "bold",
                },
                undefined,
                undefined,
              ]
            }
          >
            $
            20
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
